### ISSUE TITLE
primary key generation with 'before create' like hook

### ISSUE THREAD

**Issue by tejo-ak** [themetalfleece/neogma#9]
I want to manage id generation across different Models consistently.
Is it possible to add a new hook right before a record is saved to the database?

Thanks.

**Comment by themetalfleece** [comment]
This does sound useful. I'll get to it soon

**Comment by themetalfleece** [comment]
Published in `1.7.0`, docs can be found [here](https://themetalfleece.github.io/neogma-docs/docs/Models/Hooks).
Let me know if there is anything that's missing.

**Comment by tejo-ak** [comment]
Perfect!
Thanks a lot for considering the suggestion and implementing it.

**Comment by tejo-ak** [comment]
Hello, 
I tested the new version with the hook. It works like a charm.
I noticed that you put the hook after schema validation. 
Since the hook can change the data after being validated, the validation is too early and may not be very useful.
Also, one of the purposes of using the hook was to set a primary key (id). Therefore my data which has no id when initialized failed the validation. (I have to set my id 'required' false).

**Comment by themetalfleece** [comment]
That makes sence. Do you think that `beforeCreate` should run before the validation, or should a new hook `beforeValidate` be added? Like with [sequelize](https://sequelize.org/master/manual/hooks.html), where `beforeValidate` runs, then the instance is validated, and then `beforeCreate` runs.

**Comment by tejo-ak** [comment]
It works for me.

Making the Sequelize as a reference is a good move.

**Comment by themetalfleece** [comment]
Thanks for your input.
I tested `beforeValidate/afterValidate` hooks, and wasn't happy with the result. It added complexity, while making some things unclear. Also, I'm not in favor of mutating data in the `beforeValidate` hook.
Therefore, I implemented the original concept of `beforeCreate` running before validating the instance.

Published in `1.8.1`.

This solution should solve your issue, let me know if anything doesn't go as planned!
